File: Parallel/code/php/parallel.php
<pre>
<?php
// Copyright (c) 2014 Steve Thomas <steve AT tobtu DOT com>

function PHS($outlen, $in, $salt, $t_cost, $m_cost)
{
	$sequentialLoops = $t_cost >> 16;
	$parallelLoops   = $t_cost & 0xffff;
	if ($m_cost > 18 || $parallelLoops > 30 || $sequentialLoops > 31 || $outlen > 64)
	{
		return false;
	}
	// $sequentialLoops = 1, 2, 3, 4, 6, 8, 12, 16, ...
	if ($sequentialLoops == 0)
	{
		$sequentialLoops = 1;
	}
	else
	{
		$sequentialLoops = (3 - ($sequentialLoops & 1)) << (($sequentialLoops - 1) >> 1);
	}
	// $parallelLoops = 2, 3, 4, 6, 8, 12, 16, ...
	if ($parallelLoops == 0)
	{
		$parallelLoops = 1;
	}
	else
	{
		$parallelLoops = (3 - ($parallelLoops & 1)) << (($parallelLoops - 1) >> 1);
	}
	$parallelLoops = 3 * 5 * 128 * $parallelLoops;

	// key = SHA512(SHA512(salt) || in)
	$key = mhash(MHASH_SHA512, mhash(MHASH_SHA512, $salt) . $in);

	// Work
	for ($i = 0; $i < $sequentialLoops; $i++)
	{
		// Clear work
		$work = str_repeat("\0", 64);

		for ($j = 0; $j < $parallelLoops; $j++)
		{
			// work ^= SHA512(WRITE_BIG_ENDIAN_64(i) || WRITE_BIG_ENDIAN_64(j) || key)
			$work ^= mhash(MHASH_SHA512, pack('N4', 0, $i, 0, $j) . $key);
		}

		// Finish
		// key = truncate(SHA512(SHA512(work || key)), outlen) || zeros(64 - outlen)
		$key = substr(mhash(MHASH_SHA512, mhash(MHASH_SHA512, $work . $key)), 0, $outlen) . str_repeat("\0", 64 - $outlen);
	}

	return substr($key, 0, $outlen);
}

function testVector($p, $s, $t, $m, $ol)
{
	echo "outlen:       $ol\n";
	echo "password hex: " . bin2hex($p) . "\n";
	echo "salt hex:     " . bin2hex($s) . "\n";
	printf("t_cost:       0x%05x\n", $t);
	echo "m_cost:       0\n";
	echo "hash:         " . bin2hex(PHS($ol, $p, $s, $t, $m)) . "\n\n";
}

testVector('password', 'salt', 0x00000, 0, 64);
testVector('password', 'salt', 0x00000, 0, 32);
testVector('passwordpassword', 'salt', 0x00001, 0, 64);
testVector('password', 'saltsalt', 0x00001, 0, 64);
testVector('password', 'salt', 0x00001, 0, 64);
testVector('password', 'salt', 0x00002, 0, 64);
testVector('password', 'salt', 0x00003, 0, 64);
testVector('password', 'salt', 0x00004, 0, 64);
testVector('password', 'salt', 0x10001, 0, 64);
testVector('password', 'salt', 0x20002, 0, 64);
testVector('password', 'salt', 0x30003, 0, 64);
testVector('password', 'salt', 0x40004, 0, 64);

testVector("\xff\0", "\x80", 0x00000, 0, 64);
testVector("\xff\0", "\x80", 0x10001, 0, 64);
testVector("\xff\0", "\x80", 0x20002, 0, 64);
testVector("\xff\0", "\x80", 0x30003, 0, 64);

for ($i = 0; $i < 32; $i++)
{
	$p = '';
	for ($j = 0; $j < $i; $j++)
	{
		$p .= chr($j);
	}
	testVector($p, strrev($p), 0x00002, 0, 64);
}

File: Parallel/code/c++/parallel.cpp
// Copyright (c) 2014 Steve Thomas <steve AT tobtu DOT com>

#include <assert.h>
#include <string.h>
#include "parallel.h"
#include "sha512.h"

#define HASH_LENGTH    Sha512::HASH_LENGTH

inline uint64_t calcLoopCount(uint32_t cost)
{
	// floor((cost & 1 ? 2 : 3) * 2 ** floor((cost - 1) / 2))
	// 1, 2, 3, 4, 6, 8, 12, 16, ...
	if (cost == 0)
	{
		return 1;
	}
	return ((uint64_t) ((cost & 1) ^ 3)) << ((cost - 1) >> 1);
}

int PHS(void *out, size_t outlen, const void *in, size_t inlen, const void *salt, size_t saltlen, unsigned int t_cost, unsigned int m_cost)
{
	uint64_t key [HASH_LENGTH / sizeof(uint64_t)];
	uint64_t tmp [HASH_LENGTH / sizeof(uint64_t)];
	uint64_t work[HASH_LENGTH / sizeof(uint64_t)];
	uint64_t parallelLoops;
	uint64_t sequentialLoops;
	Sha512   sha512;

	if ((t_cost & 0xffff) > 106 || (t_cost >> 16) > 126 || outlen > HASH_LENGTH)
	{
		memset(out, 0, outlen);
		return 1;
	}

	// key = SHA512(SHA512(salt) || in)
	Sha512::hash(salt, saltlen, key);
	sha512.init();
	sha512.update(key, HASH_LENGTH);
	sha512.update(in,  inlen);
	sha512.finish(key);

	// Work
	parallelLoops = 3 * 5 * 128 * calcLoopCount(t_cost & 0xffff);
	sequentialLoops = calcLoopCount(t_cost >> 16);

	for (uint64_t i = 0; i < sequentialLoops; i++)
	{
		// Clear work
		memset(work, 0, HASH_LENGTH);

		for (uint64_t j = 0; j < parallelLoops; j++)
		{
			// work ^= SHA512(WRITE_BIG_ENDIAN_64(i) || WRITE_BIG_ENDIAN_64(j) || key)
			uint64_t tmpI = WRITE_BIG_ENDIAN_64(i);
			uint64_t tmpJ = WRITE_BIG_ENDIAN_64(j);

			sha512.init();
			sha512.update(&tmpI, sizeof(tmpI));
			sha512.update(&tmpJ, sizeof(tmpJ));
			sha512.update(key,   HASH_LENGTH);
			sha512.finish(tmp);
			for (unsigned int k = 0; k < HASH_LENGTH / sizeof(uint64_t); k++)
			{
				work[k] ^= tmp[k];
			}
		}

		// Finish
		// key = truncate(SHA512(SHA512(work || key)), outlen) || zeros(HASH_LENGTH - outlen)
		sha512.init();
		sha512.update(work, HASH_LENGTH);
		sha512.update(key,  HASH_LENGTH);
		sha512.finish(key);
		Sha512::hash(key, HASH_LENGTH, key, outlen);
		memset(((uint8_t*) key) + outlen, 0, HASH_LENGTH - outlen);
	}

	// Finish
	// out = key
	memcpy(out, key, outlen);

	// Clean up
	// TODO: find a secure wipe function
	memset(key,  0, sizeof(key));
	memset(work, 0, sizeof(key));
	memset(tmp,  0, sizeof(key));
	return 0;
}

int parallelKdf(void *out, size_t outlen, const void *in, size_t inlen, const void *salt, size_t saltlen, unsigned int t_cost, unsigned int m_cost)
{
	uint64_t key [HASH_LENGTH / sizeof(uint64_t)];
	uint64_t tmp [HASH_LENGTH / sizeof(uint64_t)];
	uint64_t work[HASH_LENGTH / sizeof(uint64_t)] = {0};
	uint64_t parallelLoops;
	Sha512   sha512;

	if (t_cost > 106 || outlen > HASH_LENGTH)
	{
		memset(out, 0, outlen);
		return 1;
	}

	// key = SHA512(SHA512(salt) || in)
	Sha512::hash(salt, saltlen, key);
	sha512.init();
	sha512.update(key, HASH_LENGTH);
	sha512.update(in,  inlen);
	sha512.finish(key);

	// Work
	parallelLoops = 3 * 5 * 128 * calcLoopCount(t_cost & 0xffff);
	for (uint64_t i = 0; i < parallelLoops; i++)
	{
		// work ^= SHA512(WRITE_BIG_ENDIAN_64(i) || key)
		uint64_t tmpI = WRITE_BIG_ENDIAN_64(i);

		sha512.init();
		sha512.update(&tmpI, sizeof(tmpI));
		sha512.update(key,   HASH_LENGTH);
		sha512.finish(tmp);
		for (unsigned int j = 0; j < HASH_LENGTH / sizeof(uint64_t); j++)
		{
			work[j] ^= tmp[j];
		}
	}

	// Finish
	// key = truncate(SHA512(SHA512(work || key)), outlen) || zeros(HASH_LENGTH - outlen)
	for (int32_t i = 0, left = (int32_t) outlen; left > 0; left -= HASH_LENGTH, i++)
	{
		uint64_t tmpI = WRITE_BIG_ENDIAN_64(i);

		// out = out || SHA512(WRITE_BIG_ENDIAN_64(i) || work || in)
		sha512.init();
		sha512.update(&tmpI, sizeof(tmpI));
		sha512.update(work,  HASH_LENGTH);
		sha512.update(in,    inlen);
		sha512.finish(out,   outlen);
		out = ((uint8_t*) out) + HASH_LENGTH;
	}

	// Clean up
	// TODO: find a secure wipe function
	memset(key,  0, sizeof(key));
	memset(work, 0, sizeof(key));
	memset(tmp,  0, sizeof(key));
	return 0;
}
